src/stirling/bpf_tools/rr/rr.cc

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "src/stirling/bpf_tools/rr/rr.h"
+
+#include <fstream>
+
+namespace px {
+namespace stirling {
+namespace bpf_tools {
+
+void BPFRecorder::WriteProto(const std::string& proto_buf_file_path) {
+  if (!recording_written_) {
+    LOG(INFO) << "Writing BPF events pb to file: " << proto_buf_file_path;
+
+    std::fstream outfile(proto_buf_file_path, std::ios::out | std::ios::trunc | std::ios::binary);
+    if (!outfile.is_open()) {
+      return;
+    }
+
+    if (!events_proto_.SerializeToOstream(&outfile)) {
+      return;
+    }
+  }
+  recording_written_ = true;
+}
+
+void BPFRecorder::RecordPerfBufferEvent(PerfBufferSpec* pb_spec, void const* const data,
+                                        const int data_size) {
+  auto event = events_proto_.add_event()->mutable_perf_buffer_event();
+  auto pb_name = event->mutable_name();
+  auto pb_data = event->mutable_data();
+
+  const std::string data_as_string(static_cast<char const*>(data), data_size);
+
+  *pb_name = pb_spec->name;
+  *pb_data = data_as_string;
+}
+
+void RecordPerfBufferEvent(void* cb_cookie, void* data, int data_size) {
+  PerfBufferSpec* pb_spec = static_cast<PerfBufferSpec*>(cb_cookie);
+  BPFRecorder* recorder = pb_spec->recorder;
+  ECHECK(recorder != nullptr);
+
+  recorder->RecordPerfBufferEvent(pb_spec, data, data_size);
+  pb_spec->probe_output_fn(pb_spec->cb_cookie, data, data_size);
+}
+
+void RecordPerfBufferLoss(void* cb_cookie, uint64_t lost) {
+  PerfBufferSpec* pb_spec = static_cast<PerfBufferSpec*>(cb_cookie);
+  pb_spec->probe_loss_fn(pb_spec->cb_cookie, lost);
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////////////////
+// Replay.
+void BPFReplayer::ReplayPerfBufferEvents(const PerfBufferSpec& perf_buffer_spec) {
+  if (PlabackComplete()) {
+    LOG_FIRST_N(INFO, 1) << "BPFReplayer::ReplayPerfBufferEvents(), plaback complete.";
+    return;
+  }
+
+  const auto n_events = events_proto_.event_size();
+
+  while (playback_event_idx_ < n_events) {
+    const auto event = events_proto_.event(playback_event_idx_);
+    if (!event.has_perf_buffer_event()) {
+      break;
+    }
```

Comment:
The recording will have different kinds of events. Therefore, "no" it is not guaranteed that the ... recording will only have perf buffer events.

We do want this to be a `break` statement so that control is returned to the containing context. For example, suppose that the socket tracer is polling perf buffers. This loop will consume all perf buffer events that match the perf buffer name, and then return control to the socket tracer. Next, the socket tracer might read a BPF map.

By breaking, we return control back to the caller when the appropriate perf buffer events are drained, i.e. so that the caller can resume and potentially read from other perf buffers or other BPF data structures.

Reply:
I added a comment for this.

Reply:
Got it. I see how that works now 👍